pkg/cloud/services/network/eips_test.go

```diff
@@ -0,0 +1,174 @@
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	tests := []struct {
+		name    string
+		expect  func(m *mock_ec2iface.MockEC2APIMockRecorder)
+		wantErr bool
+	}{
+		{
+			name: "Should return error if failed to describe IP addresses",
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeAddresses(gomock.AssignableToTypeOf(&ec2.DescribeAddressesInput{})).Return(nil, awserrors.NewFailedDependency("dependency failure"))
+			},
+			wantErr: true,
+		},
+		{
+			name: "Should ignore releasing elastic IP addresses if not found",
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeAddresses(gomock.AssignableToTypeOf(&ec2.DescribeAddressesInput{})).Return(nil, nil)
+			},
+		},
+		{
+			name: "Should return error if failed to disassociate IP address",
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeAddresses(gomock.AssignableToTypeOf(&ec2.DescribeAddressesInput{})).Return(&ec2.DescribeAddressesOutput{
+					Addresses: []*ec2.Address{
+						{
+							AssociationId: aws.String("association-id-1"),
+							PublicIp:      aws.String("public-ip"),
+							AllocationId:  aws.String("allocation-id"),
+						},
+					},
+				}, nil)
+				m.DisassociateAddress(gomock.AssignableToTypeOf(&ec2.DisassociateAddressInput{})).Return(nil, awserrors.NewFailedDependency("dependency-failure"))
+			},
+			wantErr: true,
+		},
+		{
+			name: "Should be able to release the IP address",
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeAddresses(gomock.AssignableToTypeOf(&ec2.DescribeAddressesInput{})).Return(&ec2.DescribeAddressesOutput{
+					Addresses: []*ec2.Address{
+						{
+							AssociationId: aws.String("association-id-1"),
+							PublicIp:      aws.String("public-ip"),
+							AllocationId:  aws.String("allocation-id"),
+						},
+					},
+				}, nil)
+				m.DisassociateAddress(gomock.AssignableToTypeOf(&ec2.DisassociateAddressInput{})).Return(nil, nil)
+				m.ReleaseAddress(gomock.AssignableToTypeOf(&ec2.ReleaseAddressInput{})).Return(nil, nil)
+			},
+		},
+		{
+			name: "Should retry if unable to release the IP address because of Auth Failure",
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeAddresses(gomock.AssignableToTypeOf(&ec2.DescribeAddressesInput{})).Return(&ec2.DescribeAddressesOutput{
+					Addresses: []*ec2.Address{
+						{
+							AssociationId: aws.String("association-id-1"),
+							PublicIp:      aws.String("public-ip"),
+							AllocationId:  aws.String("allocation-id"),
+						},
+					},
+				}, nil)
+				m.DisassociateAddress(gomock.AssignableToTypeOf(&ec2.DisassociateAddressInput{})).Return(nil, nil).Times(2)
+				m.ReleaseAddress(gomock.AssignableToTypeOf(&ec2.ReleaseAddressInput{})).Return(nil, awserr.New(awserrors.AuthFailure, awserrors.AuthFailure, errors.Errorf(awserrors.AuthFailure)))
+				m.ReleaseAddress(gomock.AssignableToTypeOf(&ec2.ReleaseAddressInput{})).Return(nil, nil)
```

Comment:
Why we have two different expect calls for `ReleaseAddress` here, since we have `WaitForWithRetryable` used in code and If we want to throw AuthError, wouldn't one expect call suffice?
P.S applicable to other failure cases as well.

Reply:
Actually in this test case I want to check retry happens when we receive `AuthFailure` error. So for first time I am mocking `ReleaseAddress` with AuthFailure to trigger retry. And after 1st retry I exited from retry function by returning without error from `ReleaseAddress`